src/fetch/transfer_file.py

```diff
@@ -1,51 +1,119 @@
+    @classmethod
+    def from_dict(cls, obj: dict) -> Transfer:
+        """Converts Dune data dict to object with types"""
+        token_type = TokenType.from_str(obj['token_type'])
+        token_address = obj['token_address']
+        if token_type == TokenType.NATIVE and token_address is not None:
+            raise ValueError("Native transfers must have null token_address")
+        if token_type == TokenType.ERC20 and token_address is None:
+            raise ValueError("ERC20 transfers must have valid token_address")
+
+        return cls(
+            token_type=token_type,
+            token_address=Address(
+                token_address) if token_type != TokenType.NATIVE else None,
+            receiver=Address(obj['receiver']),
+            amount=float(obj['amount'])
+        )
+
+    def add_slippage(self, slippage: Optional[SolverSlippage]):
+        if slippage is None:
+            return
+        assert self.receiver == slippage.solver_address
+        adjustment = slippage.amount_wei / 10 ** 18
+        print(
+            f"Adjusting {self.receiver} transfer by {adjustment:.5f} (slippage)"
+        )
+        new_amount = self.amount + adjustment
+        if new_amount <= 0:
+            raise ValueError(f"Invalid adjustment {self} by {adjustment}")
+        self.amount = new_amount
 
 
 def get_transfers(
         dune: DuneAnalytics,
         period_start: datetime,
         period_end: datetime
 ) -> list[Transfer]:
-    data_set = dune.fetch(
+    reimbursements_and_rewards = dune.fetch(
         query_str=dune.open_query("./queries/period_transfers.sql"),
         network=Network.MAINNET,
-        name="Period Transfers",
+        name="ETH Reimbursement & COW Rewards",
         parameters=[
             QueryParameter.date_type("StartTime", period_start),
             QueryParameter.date_type("EndTime", period_end),
         ])
-    return [
-        Transfer(
-            token_type=row['token_type'],
-            token_address=row['token_address'] or "",
-            receiver=row['receiver'],
-            amount=row['amount'],
-        )
-        for row in data_set
-    ]
+
+    negative_slippage = get_period_slippage(
+        dune, period_start, period_end, allow_positive=False
+    )
+    indexed_slippage = index_by(negative_slippage, 'solver_address')
+
+    results = []
+    for row in reimbursements_and_rewards:
+        transfer = Transfer.from_dict(row)
+        if transfer.token_type == TokenType.NATIVE:
+            slippage: SolverSlippage = indexed_slippage.get(transfer.receiver)
+            try:
+                transfer.add_slippage(slippage)
+            except ValueError as err:
+                print(
+                    f"Failed to add slippage: {err} \n"
+                    f"   Excluding eth reimbursement for solver "
+                    f"{slippage.solver_address}({slippage.solver_name})"
+                )
+                continue
+
+        results.append(transfer)
```

Comment:
In the CIP, it's written:
> a total negative slippage may be withheld from the reward

How does one do that while COWs don't have value? Sorry, its more a question rather than input :D

Reply:
I am also in favour of keeping it how it is, however it does appear the CIP specifies otherwise. Sorry I missed this.